File: Python/DataStream_ChangeResend/datastream_change_resend.py
"""
    datastream_chage_resend.py
    このサンプルはIStDataStreamのMaximum Resend Requests Per Blockを50(Default)から0に変更する方法を示します。
    * 0に変更することよりリセンド要求の上限を外れますことで、ずっと転送不安定な環境の場合ドロップ悪化する可能性はあります。

    This sample shows how to modify "Maximum Resend Requests Per Block" of DataStream from 50(default) to 0.
    * Please note that by setting this value to 0, the limitation of resend request block will be disabled.
      If you are using a unstable transferring environment, modify this value may cause multiple frame drops.
"""

import stapipy as st

number_of_images_to_grab = 100

try:
    st.initialize()

    st_system = st.create_system()

    st_device = st_system.create_first_device()

    print('Device=', st_device.info.display_name)

    st_datastream = st_device.create_datastream()
    
    # ============================================================
    # DataSteam対してのノードマップ生成
    # 注意：この設定はホストサイドの設定となります
    # Acquire NodeMap of DataStream for setting.
    # Note: This is a host side setting, not camera side.
    datastream_nodemap = st_datastream.port.nodemap
    
    # Maximum Resend Requests Per Blockのノードを取得
    # Acquire node for MaximumResendRequestsPerBlock
    max_resend_block_cnt_node = datastream_nodemap.get_node("MaximumResendRequestsPerBlock")

    # IIntタイプに変換で操作
    # Convert to IInt type for operation
    int_max_resend_block_cnt = st.PyIInteger(max_resend_block_cnt_node)
    
    # 0に設置よりリセンド要求の上限を外す
    # Set the value to 0 to disable the limitation of resend.
    int_max_resend_block_cnt.set_value(0)
    
    # ============================================================

    st_datastream.start_acquisition(number_of_images_to_grab)

    st_device.acquisition_start()

    while st_datastream.is_grabbing:
        with st_datastream.retrieve_buffer() as st_buffer:
            if st_buffer.info.is_image_present:
                st_image = st_buffer.get_image()
                print("BlockID={0} Size={1} x {2} First Byte={3}".format(
                      st_buffer.info.frame_id,
                      st_image.width, st_image.height,
                      st_image.get_image_data()[0]))
            else:
                print("Image data does not exist.")

    st_device.acquisition_stop()

    st_datastream.stop_acquisition()

except Exception as exception:
    print(exception)

File: C++/DataStream_ChangeResend/DataStream_ChangeResend.cpp
//
// DataStream_ChangeResend.cpp : 
//
// このサンプルはIStDataStreamのMaximum Resend Requests Per Blockを50(Default)から0に変更する方法を示します。
// * 0に変更することよりリセンド要求の上限を外れますことで、ずっと転送不安定な環境の場合ドロップ悪化する可能性はあります。
//
// This sample shows how to modify "Maximum Resend Requests Per Block" of DataStream from 50(default) to 0.
// * Please note that by setting this value to 0, the limitation of resend request block will be disabled.
//   If you are using a unstable transferring environment, modify this value may cause multiple frame drops.
//

#include <StApi_TL.h>

using namespace StApi;
using namespace std;
const uint64_t nCountOfImagesToGrab = 100;

int main(int /* argc */, char ** /* argv */)
{
	try
	{
		CStApiAutoInit objStApiAutoInit;
		CIStSystemPtr pIStSystem(CreateIStSystem(StSystemVendor_Sentech));
		CIStDevicePtr pIStDevice(pIStSystem->CreateFirstIStDevice());
		cout << "Device=" << pIStDevice->GetIStDeviceInfo()->GetDisplayName() << endl;

		CIStDataStreamPtr pIStDataStream(pIStDevice->CreateIStDataStream(0));

		// ==============================================================================================================
		// DataSteam対してのノードマップ生成
		// Create NodeMap for accessing DataStream.
		GenApi::CNodeMapPtr pNodeMapCameraParam(pIStDataStream->GetIStPort()->GetINodeMap());

		// Maximum Resend Requests Per Blockのノードを取得
		// Acquire Node for "MaximumResendRequestsPerBlock"
		GenApi::CIntegerPtr pIntegerResendRequests(pNodeMapCameraParam->GetNode("MaximumResendRequestsPerBlock"));
		// Maximum Resend Requests Per Blockに0を設定
		// Set MaximumResendRequestsPerBloxk to 0 (to disable the limited of resend.
		pIntegerResendRequests->SetValue(0);

		// ==============================================================================================================

		pIStDataStream->StartAcquisition(nCountOfImagesToGrab);

		pIStDevice->AcquisitionStart();

		while (pIStDataStream->IsGrabbing())
		{
			CIStStreamBufferPtr pIStStreamBuffer(pIStDataStream->RetrieveBuffer(5000));

			if (pIStStreamBuffer->GetIStStreamBufferInfo()->IsImagePresent())
			{
				IStImage *pIStImage = pIStStreamBuffer->GetIStImage();

				cout << "BlockId=" << pIStStreamBuffer->GetIStStreamBufferInfo()->GetFrameID()
					<< " Size:" << pIStImage->GetImageWidth() << " x " << pIStImage->GetImageHeight()
					<< " First byte =" << (uint32_t)*(uint8_t*)pIStImage->GetImageBuffer() << endl;
			}
			else
			{
				cout << "Image data does not exist" << endl;
			}
		}

		pIStDevice->AcquisitionStop();

		pIStDataStream->StopAcquisition();
	}
	catch (const GenICam::GenericException &e)
	{
		cerr << endl << "An exception occurred." << endl << e.GetDescription() << endl;
	}

	cout << endl << "Press Enter to exit." << endl;
	while (cin.get() != '\n');

	return(0);
}